hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngradeUtils.java

```diff
@@ -180,4 +191,165 @@ public static long convertCompletionTimeToEpoch(HoodieInstant instant) {
+      // TODO: Check if we should hardcode disable rollback using markers.
+      //       With changes in https://github.com/apache/hudi/pull/12206, we may not need to do that.
+      rollbackWriteConfig.setValue(HoodieWriteConfig.ROLLBACK_USING_MARKERS_ENABLE.key(), String.valueOf(config.shouldRollbackUsingMarkers()));
+      // set compaction configs
+      if (shouldCompact) {
+        rollbackWriteConfig.setValue(HoodieCompactionConfig.INLINE_COMPACT.key(), "true");
+        rollbackWriteConfig.setValue(HoodieCompactionConfig.INLINE_COMPACT_NUM_DELTA_COMMITS.key(), "1");
+        rollbackWriteConfig.setValue(HoodieCompactionConfig.INLINE_COMPACT_TRIGGER_STRATEGY.key(), CompactionTriggerStrategy.NUM_COMMITS.name());
+        rollbackWriteConfig.setValue(HoodieCompactionConfig.COMPACTION_STRATEGY.key(), UnBoundedCompactionStrategy.class.getName());
+      } else {
+        rollbackWriteConfig.setValue(HoodieCompactionConfig.INLINE_COMPACT.key(), "false");
+      }
+      // TODO: Check if we need to disable metadata table.
+      rollbackWriteConfig.setValue(HoodieMetadataConfig.ENABLE.key(), String.valueOf(config.isMetadataTableEnabled()));
+      // TODO: This util method is used in both upgrade and downgrade.
+      //       Check if we need to instantiate write client with correct table version or is that handled internally.
+      try (BaseHoodieWriteClient writeClient = upgradeDowngradeHelper.getWriteClient(rollbackWriteConfig, context)) {
+        writeClient.rollbackFailedWrites();
+        HoodieTableMetaClient.reload(table.getMetaClient());
+        if (shouldCompact) {
+          Option<String> compactionInstantOpt = writeClient.scheduleCompaction(Option.empty());
+          if (compactionInstantOpt.isPresent()) {
+            writeClient.compact(compactionInstantOpt.get());
+          }
+        }
+      }
+    } catch (Exception e) {
+      throw new HoodieException(e);
+    }
+  }
+
+  static void upgradeToLSMTimeline(HoodieTable table, HoodieEngineContext engineContext, HoodieWriteConfig config) {
+    table.getMetaClient().getTableConfig().getTimelineLayoutVersion().ifPresent(
+        timelineLayoutVersion -> ValidationUtils.checkState(TimelineLayoutVersion.LAYOUT_VERSION_1.equals(timelineLayoutVersion),
+            "Upgrade to LSM timeline is only supported for layout version 1. Given version: " + timelineLayoutVersion));
+    HoodieArchivedTimeline archivedTimeline = table.getMetaClient().getArchivedTimeline();
+    if (archivedTimeline.getInstants().isEmpty()) {
+      return;
+    }
+    Consumer<Exception> exceptionHandler = e -> {
+      if (config.isFailOnTimelineArchivingEnabled()) {
+        throw new HoodieException(e);
+      }
+    };
+    try {
+      HoodieTimelineArchiver archiver = TimelineArchivers.getInstance(TimelineLayoutVersion.LAYOUT_VERSION_2, config, table);
+      archiver.archiveInstants(engineContext, archivedTimeline.getInstants(), true);
```

Comment:
I don't think we can read the legady archived instants in one shot. The right manner is reading them in mini-batchs then flush as the new LSM timeline.

Reply:
And be caution that the legacy archived instants can be large amount, just reading the latest log should be enough for the new file slicing.

Reply:
read the legacy archived instants as a iterator and flush as a accumulated mini-batch.